datarace on pebble with statedb (testing)

```
WARNING: DATA RACE
Write at 0x00c0009ac038 by goroutine 401:
  github.com/cockroachdb/pebble.newIndexedBatch()
      /home/runner/go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20230620232302-06034ff014e0/batch.go:371 +0x1a4
  github.com/cockroachdb/pebble.(*DB).NewIndexedBatch()
      /home/runner/go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20230620232302-06034ff014e0/db.go:1424 +0x7a
  go.vocdoni.io/dvote/db/pebbledb.(*PebbleDB).WriteTx()
      /home/runner/work/vocdoni-node/vocdoni-node/db/pebbledb/pebledb.go:138 +0x25
  go.vocdoni.io/dvote/statedb.(*StateDB).BeginTx()
      /home/runner/work/vocdoni-node/vocdoni-node/statedb/statedb.go:301 +0xc5
  go.vocdoni.io/dvote/vochain/state.NewState()
      /home/runner/work/vocdoni-node/vocdoni-node/vochain/state/state.go:124 +0x311
  go.vocdoni.io/dvote/test/testcommon.NewVochainState()
      /home/runner/work/vocdoni-node/vocdoni-node/test/testcommon/vochain.go:86 +0x4f
  go.vocdoni.io/dvote/test/testcommon.NewVochainStateWithProcess()
      /home/runner/work/vocdoni-node/vocdoni-node/test/testcommon/vochain.go:129 +0x34
  go.vocdoni.io/dvote/test.TestCountVotes()
      /home/runner/work/vocdoni-node/vocdoni-node/test/statedb_test.go:112 +0x4f
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x44

Previous read at 0x00c0009ac038 by goroutine 395:
  github.com/cockroachdb/pebble.(*Batch).release()
      /home/runner/go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20230620232302-06034ff014e0/batch.go:388 +0x3c
  github.com/cockroachdb/pebble.(*Batch).Close()
      /home/runner/go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20230620232302-06034ff014e0/batch.go:1283 +0x26
  go.vocdoni.io/dvote/db/pebbledb.WriteTx.Discard()
      /home/runner/work/vocdoni-node/vocdoni-node/db/pebbledb/pebledb.go:96 +0xe
  go.vocdoni.io/dvote/db/prefixeddb.(*PrefixedWriteTx).Discard()
      /home/runner/work/vocdoni-node/vocdoni-node/db/prefixeddb/prefixeddb.go:129 +0x42
  go.vocdoni.io/dvote/vochain/state.(*State).Close()
      /home/runner/work/vocdoni-node/vocdoni-node/vochain/state/state.go:470 +0xc7
  go.vocdoni.io/dvote/test/testcommon.NewVochainState.func1()
      /home/runner/work/vocdoni-node/vocdoni-node/test/testcommon/vochain.go:90 +0x2e
  testing.(*common).Cleanup.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1169 +0x182
  testing.(*common).runCleanup()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1[347](https://github.com/vocdoni/vocdoni-node/actions/runs/6470977793/job/17568382973#step:5:348) +0x1c2
  testing.tRunner.func2()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1589 +0x50
  runtime.deferreturn()
      /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:477 +0x30
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x44

Goroutine 401 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x82a
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:2054 +0x84
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1595 +0x238
  testing.runTests()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:2052 +0x896
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1925 +0xb57
  main.main()
      _testmain.go:103 +0x2e4

Goroutine 395 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x82a
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:2054 +0x84
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1595 +0x238
  testing.runTests()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:2052 +0x896
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1925 +0xb57
  main.main()
      _testmain.go:103 +0x2e4
==================
--- FAIL: TestCountVotes (0.03s)
    testing.go:1465: race detected during execution of test
--- FAIL: TestAddValidator (0.04s)
    testing.go:1465: race detected during execution of test
FAIL
```

--- a/db/interface.go
+++ b/db/interface.go
@@ -73,12 +73,12 @@ type WriteTx interface {
 	// Apply internally needs type assertions, revisit this once generics
 	// are ready.
 	Apply(WriteTx) error
-	// Commit commits the transaction into the db
+	// Commit commits the transaction into the db.
+	// Calling Commit more than once, or after Discard, is an error.
 	Commit() error
-	// Discard discards the transaction. This method can be called always,
-	// even if previously the Tx has been committed (for the WriteTx case).
-	// So it's a good practice to `defer tx.Discard()` just after creating
-	// the tx.
+	// Discard releases the transaction's resources as they don't need to be committed.
+	// This method can be safely called after any previous Commit or Discard call,
+	// for the sake of allowing deferred Discard calls.
 	Discard()
 }
 

--- a/db/pebbledb/pebledb.go
+++ b/db/pebbledb/pebledb.go
@@ -2,6 +2,7 @@ package pebbledb
 
 import (
 	"errors"
+	"fmt"
 	"os"
 
 	"github.com/cockroachdb/pebble"
@@ -61,39 +62,53 @@ func iterate(reader pebble.Reader, prefix []byte, callback func(k, v []byte) boo
 }
 
 // Get implements the db.WriteTx.Get interface method
-func (tx WriteTx) Get(k []byte) ([]byte, error) {
+func (tx *WriteTx) Get(k []byte) ([]byte, error) {
 	return get(tx.batch, k)
 }
 
-func (tx WriteTx) Iterate(prefix []byte, callback func(k, v []byte) bool) (err error) {
+func (tx *WriteTx) Iterate(prefix []byte, callback func(k, v []byte) bool) (err error) {
 	return iterate(tx.batch, prefix, callback)
 }
 
 // Set implements the db.WriteTx.Set interface method
-func (tx WriteTx) Set(k, v []byte) error {
+func (tx *WriteTx) Set(k, v []byte) error {
 	return tx.batch.Set(k, v, nil)
 }
 
 // Delete implements the db.WriteTx.Delete interface method
-func (tx WriteTx) Delete(k []byte) error {
+func (tx *WriteTx) Delete(k []byte) error {
 	return tx.batch.Delete(k, nil)
 }
 
 // Apply implements the db.WriteTx.Apply interface method
-func (tx WriteTx) Apply(other db.WriteTx) (err error) {
-	otherPebble := db.UnwrapWriteTx(other).(WriteTx)
+func (tx *WriteTx) Apply(other db.WriteTx) (err error) {
+	otherPebble := db.UnwrapWriteTx(other).(*WriteTx)
 	return tx.batch.Apply(otherPebble.batch, nil)
 }
 
 // Commit implements the db.WriteTx.Commit interface method
-func (tx WriteTx) Commit() error {
-	return tx.batch.Commit(nil)
+func (tx *WriteTx) Commit() error {
+	if tx.batch == nil {
+		return fmt.Errorf("cannot commit pebble tx: already committed or discarded")
+	}
+	err := tx.batch.Commit(nil)
+	tx.batch = nil
+	return err
 }
 
 // Discard implements the db.WriteTx.Discard interface method
-func (tx WriteTx) Discard() {
+func (tx *WriteTx) Discard() {
+	if tx.batch == nil {
+		// Silently allow discarding twice or after a commit,
+		// since it can help for the sake of defers.
+		// Note that upstream pebbledb doesn't really allow calling Close twice;
+		// the object gets put in a shared pool, so a later call to Close races
+		// with another goroutine having already reused the same object from the pool.
+		return
+	}
 	// Close returns an error, but here in the Discard context is omitted
 	tx.batch.Close()
+	tx.batch = nil
 }
 
 // PebbleDB implements db.Database interface
@@ -134,7 +149,7 @@ func (db *PebbleDB) Get(k []byte) ([]byte, error) {
 
 // WriteTx returns a db.WriteTx
 func (db *PebbleDB) WriteTx() db.WriteTx {
-	return WriteTx{
+	return &WriteTx{
 		batch: db.db.NewIndexedBatch(),
 	}
 }

